internal/mtail/unix_socket_export_integration_test.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+// This file is available under the Apache license.
+// +build integration
+
+package mtail_test
+
+import (
+	"fmt"
+	"path"
+	"testing"
+	"time"
+	"net"
+	"net/http"
+	"os"
+	"runtime"
+	"context"
+	"bytes"
+	"encoding/json"
+
+	"github.com/golang/glog"
+	"github.com/google/mtail/internal/mtail"
+	"github.com/google/mtail/internal/testutil"
+	"github.com/google/mtail/internal/metrics"
+	"github.com/google/mtail/internal/watcher"
+)
+
+// makeServer makes a new Server for use in tests, but does not start
+// the server.  It returns the server, or any errors the new server creates.
+func makeServer(tb testing.TB, pollInterval time.Duration, enableFsNotify bool, options ...func(*mtail.Server) error) (*mtail.Server, error) {
+	tb.Helper()
+	w, err := watcher.NewLogWatcher(pollInterval, enableFsNotify)
+	if err != nil {
+		tb.Fatal(err)
+	}
+
+	return mtail.New(metrics.NewStore(), w, options...)
+}
+
+
+// startUNIXSocketServer creates a new Server serving through a UNIX
+// socket and starts it running. It returns the server, and a cleanup function.
+func startUNIXSocketServer(tb testing.TB, pollInterval time.Duration, enableFsNotify bool, options ...func(*mtail.Server) error) (*mtail.Server, func()) {
+	tb.Helper()
+
+	unixSocket := "/var/run/mtail_test.socket"
+	options = append(options, mtail.BindUnixSocket("/var/run/mtail_test.socket"))
+
+	m, err := makeServer(tb, pollInterval, enableFsNotify, options...)
+	if err != nil {
+		tb.Fatal(err)
+	}
+
+	errc := make(chan error, 1)
+	go func() {
+		err := m.Run()
+		errc <- err
+	}()
+
+	glog.Infof("check that server is listening")
+
+	addr, err := net.ResolveUnixAddr("unix", unixSocket)
+
+	if err != nil {
+		tb.Fatal(err)
+	}
+	_, err = net.DialUnix("unix", nil, addr)
+
+	if err != nil {
+		tb.Fatal(err)
+	}
+
+	return m, func() {
+		err := m.Close()
+		if err != nil {
+			tb.Fatal(err)
+		}
+		if err != nil {
+			tb.Fatal(err)
+		}
+		select {
+		case err = <-errc:
+		case <-time.After(5 * time.Second):
+			buf := make([]byte, 1<<16)
+			n := runtime.Stack(buf, true)
+			fmt.Fprintf(os.Stderr, "%s", buf[0:n])
+			tb.Fatal("timeout waiting for shutdown")
+		}
+
+		if err != nil {
+			tb.Fatal(err)
+		}
+	}
+}
+
+
+// getMetricFromUNIXSocket fetches the name metrics from the Server serving through
+// the UNIX socket sockPath, and returns the value of one named name.  Callers are
+// responsible for type assertions on the returned value.
+func getMetricFromUNIXSocket(tb testing.TB, sockPath, name string) interface{} {
```

Comment:
I think this should be a method on the *TestServer so that we can abstract away the `http.Transport`.  But all I'll ask is that you write that down as a TODO here.  Thanks!